File: baekjoon/Python/백트래킹/15649.py
def solve(depth, N, M, out):
    if depth == M: # 탈출 조건
        print(" ".join(map(str, out)))
        return
    for i in range(len(visited)):
        if not visited[i]:
            visited[i] = True
            out.append(i+1)
            solve(depth+1, N, M, out)
            #visited[i] = False
            out.pop()
if __name__ == "__main__":
    N, M = map(int, input().split())
    visited = [False]*N
    out = []
    solve(0, N, M, out)

File: BOJ/단계별 문제/14. 백트래킹/15649.cpp
#include <stdio.h>

char a[16] = "               ";
int n, m, visited[8];

void dfs(int depth);

int main(void) {
    scanf("%d%d", &n, &m);
    m += m;
    a[m] = 0;
    dfs(0);
}

void dfs(int depth) {
    if (depth == m) {
        printf("%s\n", a);
        return;
    }
    for (int i=0; i<n; i++) {
        if (!visited[i]) {
            a[depth] = i + '1';
            visited[i] = 1;
            dfs(depth+2);
            visited[i] = 0;
        }
    }
}